### ISSUE TITLE
Consider specialcasing struct keys that fit into int/long/nint

### ISSUE THREAD

**Issue by VSadov** [VSadov/NonBlocking#16]
Re: https://github.com/VSadov/NonBlocking/pull/13#issuecomment-1155506457

Such keys could use the same implementation as primitives 
The question is mostly of complexity vs. how common these cases are.

**Comment by neon-sunset** [comment]
To be honest, it doesn't seem like a frequent case except for enums, in which case it would be reasonable to assume that they would get treated as their underlying primitive but in reality they wouldn't.

The main motivation is it's a nice to have feature in the case of `ValueTuple`s which I'm using here: https://github.com/neon-sunset/fast-cache/blob/main/src/FastCache.Cached/Cached.TryGet.cs#L14
ValueTuples in such scenario seem to produce okay-ish codegen and perf (vs record structs or structs with hand-rolled hashcode/comparer).
However, they cannot compete with codegen size and `4.5ns` of minimum latency (all code/data is in L1/L2 cache) achievable for `nint` \ `long` and take quite a hit at `~30ns` because they use `DictionaryImplBoxed` when they don't have to.

**Comment by neon-sunset** [comment]
If you're okay with this change, I'll go ahead with impl.

**Comment by VSadov** [comment]
The primary gain will be from storing keys, that can be updated atomically, inline - no boxing, no indirection. 

My only worry if implementation is too complex for the gains. The `if` ladder in the dictionary constructor is already a bit scary.

https://github.com/VSadov/NonBlocking/blob/ad40be991a54d37a5df3de34a82011c22fdc0c45/src/NonBlocking/ConcurrentDictionary/ConcurrentDictionary.cs#L178

Maybe it can stay mostly the same if switched to check sizes like

`if (sizeof(TKey) == sizeof(int) && comparer == null))`

**Comment by neon-sunset** [comment]
Yes, that's what I was thinking - coalesce implementations for fitting keys into int-sized and long-sized and then let JIT optimize out checks like `entryKey == providedKey || (!s_isPrimitive && Unsafe.As(ref entryKey) == Unsafe.As(ref providedKey))`.

**Comment by VSadov** [comment]
Runtime does somewhat similar optimization in `ValueType.Equals`. In some cases it simply compares all bits in the instance.

That is not always possible and keyed off `CanCompareBits` results.
https://github.com/dotnet/runtime/blob/d6e8686e566efbf9e921581abbfb2dc114369168/src/coreclr/System.Private.CoreLib/src/System/ValueType.cs#L41

The implementation of `CanCompareBits` in native runtime checks for two conditions:
https://github.com/dotnet/runtime/blob/021dc5f0ac9509c03ae0fa330d6fe0305d8492d2/src/coreclr/vm/comutilnative.cpp#L1616
- there are no managed references in the instance. 
This is easy to check via `IsReferenceOrContainsReferences<T>`
- the instance is tightly packed (i.e. there are no gaps between fields in the instance layout)
This is more problematic. There is no public API for this and without that figuring `IsTightlyPacked` would be difficult/slow.

The optimization can work for enums, but general case may be hard.
